src/chainparamsbase.cpp

```diff
@@ -94,22 +93,26 @@ CBaseChainParams::Network NetworkIdFromCommandLine()
     bool fTestNet = GetBoolArg("-testnet", false);
 
     if (fTestNet && fRegTest)
-        return CBaseChainParams::MAX_NETWORK_TYPES;
+        throw std::runtime_error("Invalid combination of -regtest and -testnet.\n");
     if (fRegTest)
         return CBaseChainParams::REGTEST;
     if (fTestNet)
         return CBaseChainParams::TESTNET;
-    return CBaseChainParams::MAIN;
+
+    std::string chain = GetArg("-chain", "main");
+    if (chain == "main")
+        return CBaseChainParams::MAIN;
```

Comment:
Well, we can get rid of the enumeration even if the strings aren't declared as constants. I won't opposed to it but working at big java projects were those type of constants are abundant I have always wondered if the gain in readability that CBaseChainParams::TESTNET provides compared to "testnet".
Yes, yes, developers could type "testtnet" and have some fun getting an "unkown chain 'testtnet'" exception...

Personally I like the Params(type) or Params(std::string) factory more than Params() in combination with SelectParams(). I mean, this is closer to what I would do: https://github.com/jtimon/bitcoin/commit/d135001a0a8ae9536553d49b1ac8a3c2e0643e41#diff-d22bc3e058f8982972e2eb381a1df668R34
In fact my preference when it comes to creating polymorphic objects would be for the factory function to return a pointer to a newly created object that the caller has to delete, a good-old factory without singleton [like Params()] or per-type singleton [like Params(std::string)], but that's more than we need. 
I was scared that moving from enum to string was going to be something people wouldn't like, but if that's not the case I can definitely improve this.
The remaining question is, do we really need the string constants when you will get an exception right in the face if you don't use the correct string?
In my opinion, "the same string you use in command line and it's documented" would be ideal.